independent-publisher-connectors/Infura Ethereum/README.md

```diff
@@ -0,0 +1,33 @@
+# Infura Ethereum Connector
+
+Infura provides services to access the Ethereum Blockchain without your own Ethereum Node. This connector allows you to interact with the Ethereum Blockchain JSON-RPC API through Infura. Common use cases include checking your Ethereum account's (Wallet) balance, current gas fees, or information on the current block.
+
+## Prerequisites
+
+* An Infura [Free Plan](https://infura.io/pricing) (or higher)
+* An Infura Ethereum Project
+* The Project Id of the Infura Ethereum Project
+
+## Supported Operations
+
+The connector supports the following operations:
+
+|Operation Id  |Name  |Description  |Support  |
+|---------|---------|---------|---------|
+|eth_getBlockNumber|Block Number|Returns the number of most recent block.|✅|
+|eth_getBalance|Balance|Returns the balance of the account of given address.|✅|
+|eth_gasPrice|Gas Price|Returns the current price per gas in wei.|✅|
+
+## Obtaining Credentials
+
+The Infura Ethereum Connector uses the Infura Project Id to communicate with the JSON-RPC endpoint. See the **Getting Started** section on how to get a new Project Id.
+The **Project Id** is considered a secret and securely stored inside the Power Platform connection. Do not share the Project Id.
```

Comment:
Interesting. The Project Id is supposed to be secret, but it is sent a path parameter.

Reply:
Hello @schabungbam, 

I'd like to have a discussion with you on the Project Id. The Infura API docs state the following:
> By default, the Project Secret is optional. If you know you will only be interacting with Infura APIs from a secure environment (e.g. your backend web application), you can toggle the Private Secret Required setting in your Project's configuration page under the Security section.

I think this discussion is as old as the web, but can we assume that a `securestring` parameter inside the Power Platform is never exposed externally, even when used in the path parameter? In other words, is the platform considered a 'secure environment'? 

Having said that, the Infura API does support basic auth as additional protection. However, since basic auth is just a hashed authorization token, this doesn't add much to the security if we consider the Power Platform a 'client' system instead of a 'secure environment'.

In the effort to make this as secure as possible, I tried to add basic auth to the connector (in addition to the project id), and I failed. The Infura API has a very unfortunate implementation of basic auth and expects an empty username:

`curl --user :YOUR-PROJECT-SECRET https://<network>.infura.io/v3/YOUR-PROJECT-ID`
> Note that the user field of the request is left blank and the project id is passed in the URL.

This scenario seems to be unsupported by custom connectors since the username can't be empty in the UI when setting the connector to use basic auth. 

I tried to tweak the apiProperties.json to make this work, but I've failed. Here's my approach:

```json
"connectionParameters": {
      "username": {
        "type": "string",
        "defaultValue": "",
        "uiDefinition": {
          "displayName": "Username",
          "description": "The Username for this api",
          "tooltip": "Provide the Username",
          "constraints": {
            "tabIndex": 2,
            "clearText": true,
            "required": "false",
            "hidden": "true"
          }
        }
      },
      "password": {
        "type": "securestring",
        "uiDefinition": {
          "displayName": "Project Secret",
          "description": "The Project Secret for this api",
          "tooltip": "Provide the Project Secret",
          "constraints": {
            "tabIndex": 3,
            "clearText": false,
            "required": "true"
          }
        }
      },
      "projectId": {
        "type": "securestring",
        "uiDefinition": {
          "constraints": {
            "required": "true",
            "tabIndex": 4,
            "clearText": false
          },
          "description": "The Id of your Infura Ethereum Project.",
          "displayName": "Project Id",
          "tooltip": "The Project Id can be found on the Infura Dashboard. Select your Ethereum Project and check out the KEYS section."
        }
      }
    }
```
As you can see, I attempted to make the username for basic auth to be non-mandatory and set an empty string as the default value. Unfortunately, while the resulting UI looks promising, it doesn't generate the correct hash for basic auth when testing the connector. So, unfortunately, I don't have any insights into what is happening under the hood.

![image](https://user-images.githubusercontent.com/12109119/152643141-15d60ded-90bf-4be9-93e7-5759eff2518a.png)

Also, the `hidden: "true"` attribute doesn't work, which would improve the UX if the approach overall were functional.

I favor going with just the project id, even when used in the path (assuming the Power Platform to be a 'secure environment'). 

Would you mind having a look? Are you part of the product group, and could this be a feature request or considered a 'known limitation', or should I report this to @natalie-pienkowska?

Reply:
I've added a **Securing Infura** section to the README.md. It should make the solution secure, even without basic auth. Still, I'd love to follow up on the discussion about this scenario.